lib/Dialect/AIG/Analysis/LongestPathAnalysis.cpp

```diff
@@ -172,47 +262,1100 @@ struct llvm::DenseMapInfo<Object> {
+    if (*leader != valueAndBitPos) {
+      // If this is not the leader, then use the leader.
+      auto root = ecMap.at(*leader);
+      return getResults(root.first, root.second);
+    }
+  }
+
+  auto it = cachedResults.find(valueAndBitPos);
+  // If not found, then consider it to be a constant.
+  if (it == cachedResults.end())
+    return {};
+  return it->second;
+}
+
+void LocalVisitor::putUnclosedResult(const Object &object, int64_t delay,
+                                     llvm::ImmutableList<DebugPoint> history,
+                                     ObjectToMaxDistance &objectToMaxDistance) {
+  auto &slot = objectToMaxDistance[object];
+  if (slot.first >= delay && delay != 0)
+    return;
+  slot = {delay, history};
+}
+
+void LocalVisitor::waitUntilDone() const {
+  // Wait for the thread to finish.
+  while (!done.load())
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+}
+
+LogicalResult LocalVisitor::markRegFanOut(Value fanOut, Value start,
+                                          Value reset, Value resetValue,
+                                          Value enable) {
+  auto bitWidth = getBitWidth(fanOut);
+  auto record = [&](size_t fanOutBitPos, Value value, size_t bitPos) {
+    auto result = getOrComputeResults(value, bitPos);
+    if (failed(result))
+      return failure();
+    for (auto &path : *result) {
+      if (auto blockArg = dyn_cast<BlockArgument>(path.fanIn.value)) {
+        // Not closed.
+        putUnclosedResult({{}, fanOut, fanOutBitPos}, path.delay, path.history,
+                          fromInputPortToFanOut[{blockArg, path.fanIn.bitPos}]);
+      } else {
+        // If the fanIn is not a port, record to the results.
+        fanOutResults[{{}, fanOut, fanOutBitPos}].push_back(path);
+      }
+    }
+    return success();
+  };
+
+  // Get paths for each bit, and record them.
+  for (size_t i = 0, e = bitWidth; i < e; ++i) {
+    if (failed(record(i, start, i)))
+      return failure();
+  }
+
+  // Edge from a reg to reset/resetValue.
+  if (reset)
+    for (size_t i = 0, e = bitWidth; i < e; ++i) {
+      if (failed(record(i, reset, 0)) || failed(record(i, resetValue, i)))
+        return failure();
+    }
+
+  // Edge from a reg to enable signal.
+  if (enable)
+    for (size_t i = 0, e = bitWidth; i < e; ++i) {
+      if (failed(record(i, enable, 0)))
+        return failure();
+    }
+
+  return success();
+}
+
+LogicalResult LocalVisitor::markEquivalent(Value from, size_t fromBitPos,
+                                           Value to, size_t toBitPos,
+                                           SmallVectorImpl<OpenPath> &results) {
+  auto leader = ec.getOrInsertLeaderValue({to, toBitPos});
+  auto &slot = ecMap[leader];
+  if (!slot.first)
+    slot = {to, toBitPos};
+  // Merge classes, and visit the leader.
+  auto newLeader = ec.unionSets({to, toBitPos}, {from, fromBitPos});
+  assert(leader == *newLeader);
+  return visitValue(to, toBitPos, results);
```

Comment:
Really nice idea to get rid of aliases in the IR and boil things down to a single computed value regardless of how many wires and bit slices you have of that value :100: